### ISSUE TITLE
[Contract Upgrade] Bad Token Detector

### ISSUE THREAD

**Issue by bh2smith** [gnosis/gp-v2-services#903]
This is still somewhat of an open question, but for example, we will want to prevent deflationary tokens from being traded via the Balancer vault option, since such trades may pass simulation - but are known to still be a problem.

TODO - come up with other examples where the token detector needs to be adapted according to the new order model.

**Comment by fleupold** [comment]
Could you elaborate on the example how a deflationary token would be problematic?

**Comment by vkgnosis** [comment]
The current bad token detector detects deflationary as bad afaik. I don't think this would change with the contracts upgrade since the the bad token detector doesn't use the gpv2 contracts anyway.

**Comment by bh2smith** [comment]
> The current bad token detector detects deflationary as bad afaik. I don't think this would change with the contracts upgrade since the the bad token detector doesn't use the gpv2 contracts anyway.

True, but we do still need to ponder some other examples as part of the issues TODO

**Comment by bh2smith** [comment]
> Could you elaborate on the example how a deflationary token would be problematic?

From my understanding, if we weren't to exclude these deflationary tokens from the start an order involving this token with `VaultInternal` as and token source and destination would pass the simulation. This is because there are no actual erc20 transfers performed during such a trade (when updating internal balances), it is only at the time of exiting the funds that something might break. Of course BalancerV2 doesn't support these tokens either --- as far as I know so it is unlikely that there will be any pools involving them.

**Comment by vkgnosis** [comment]
The bad token detector currently does not know about balancer vaults. So if a token is deflationary it will simulate a normal erc20 transfer and see that some balance is lost. When someone tries to place an order with such a token it would be denied by the api. Based on my understanding this whole process doesn't change with the contracts upgrade.

**Comment by bh2smith** [comment]
So then, the main purpose of this issue is to get the token detector to work also with the balancer vault

**Comment by nlordell** [comment]
So the initial concern I had was more around the bad token detector if we were to use baseline solutions to detect bad tokens. In that case, a baseline solver *could* compute a solution that does a direct Balancer swap with internal balances on a bad token that would succeed. The issue then is that the settlement contract would be stuck with some bad tokens as internal Balancer vault balances that it would not be able to take out.

However, if we are assuming that for the initial SC update that we do not need to support internal/external Vault balances, then this issue can be moved out of the milestone.
